src/Components/Components/src/ResourceAssetCollection.cs

```diff
@@ -0,0 +1,63 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+
+namespace Microsoft.AspNetCore.Components;
+
+/// <summary>
+/// Describes a mapping of static assets to their corresponding unique URLs.
+/// </summary>
+public sealed class ResourceAssetCollection : IReadOnlyList<ResourceAsset>
+{
+    /// <summary>
+    /// An empty <see cref="ResourceAssetCollection"/>.
+    /// </summary>
+    public static readonly ResourceAssetCollection Empty = new([]);
+
+    private readonly Dictionary<string, ResourceAsset> _uniqueUrlMappings;
```

Comment:
I'd be curious if it's worth considering using a `FrozenDictionary` here if we want lookups to be as fast as possible.